File: pms-fullstack/backend/db_setup.py
from sqlmodel import SQLModel, Session, select
from database import engine
from models import Faculty, Project, Student, Skill, StudentSkill, ProjectAllocation, Department, User
from datetime import date

def init_db():
    SQLModel.metadata.drop_all(engine) # Optional: Reset DB for clean state
    SQLModel.metadata.create_all(engine)
    print("Tables created successfully.")

    with Session(engine) as session:
        if session.exec(select(Department)).first():
            print("Data exists.")
            return

        # 1. Add Departments
        d1 = Department(department_name="Computer Science")
        d2 = Department(department_name="Information Technology")
        session.add(d1)
        session.add(d2)
        session.commit()

        # 2. Add Faculty (Guides)
        f1 = Faculty(first_name="Alice", last_name="Wong", email="alice@uni.edu", designation="Professor", department=d1)
        f2 = Faculty(first_name="Bob", last_name="Green", email="bob@uni.edu", designation="Asst. Professor", department=d2)
        session.add(f1)
        session.add(f2)
        session.commit()

        # 3. Add Projects
        p1 = Project(project_name="ML Data Analysis", description="COVID Trends", status="In Progress", start_date=date(2023, 1, 15), guide=f1)
        p2 = Project(project_name="Library System", description="Book tracking", status="Planned", guide=f2)
        session.add(p1)
        session.add(p2)

        # 4. Add Students
        s1 = Student(first_name="John", last_name="Doe", email="john@uni.edu", phone_number="1234567890", department=d1)
        s2 = Student(first_name="Jane", last_name="Smith", email="jane@uni.edu", phone_number="0987654321", department=d2)
        s3 = Student(first_name="Mike", last_name="Ross", email="mike@uni.edu", department=d1)
        session.add(s1)
        session.add(s2)
        session.add(s3)
        
        # 5. Add Skills
        sk1 = Skill(skill_name="Python")
        sk2 = Skill(skill_name="SQL")
        session.add(sk1)
        session.add(sk2)
        
        session.commit()

        # 6. Assign Skills & Projects
        session.add(StudentSkill(student_id=s1.student_id, skill_id=sk1.skill_id, proficiency_level="Expert"))
        session.add(ProjectAllocation(project_id=p1.project_id, student_id=s1.student_id))

        session.commit()
        print("Sample data inserted.")

        # Admin
        session.add(User(email="admin@uni.edu", password="admin", role="admin"))

        # Faculty Users (Alice as Head, Bob as Guide)
        session.add(User(email="alice@uni.edu", password="1234", role="head"))
        session.add(User(email="bob@uni.edu", password="1234", role="guide"))

        # Student Users
        session.add(User(email="john@uni.edu", password="1234", role="student"))
        session.add(User(email="jane@uni.edu", password="1234", role="student"))
        session.add(User(email="mike@uni.edu", password="1234", role="student"))

        session.commit()
        print("Data initialized.")

if __name__ == "__main__":
    init_db()

File: pms-fullstack/backend/db_setup.js
const sqlite3 = require('sqlite3').verbose();

// Connect to (or create) the database file
const db = new sqlite3.Database('./pms.db', (err) => {
    if (err) {
        console.error('Error opening database:', err.message);
    } else {
        console.log('Connected to the SQLite PMS database.');
        initDatabase();
    }
});

function initDatabase() {
    db.serialize(() => {
        // 1. Drop tables if they exist (to ensure a clean slate for testing)
        db.run("DROP TABLE IF EXISTS ProjectAllocation");
        db.run("DROP TABLE IF EXISTS Student_Skills");
        db.run("DROP TABLE IF EXISTS Skills");
        db.run("DROP TABLE IF EXISTS Students");
        db.run("DROP TABLE IF EXISTS Projects");
        db.run("DROP TABLE IF EXISTS ProjectGuides");

        // 2. Create Tables (DDL)
        
        // Faculty / Guides Table
        db.run(`CREATE TABLE ProjectGuides (
            faculty_id INTEGER PRIMARY KEY,
            first_name TEXT NOT NULL,
            last_name TEXT NOT NULL,
            email TEXT UNIQUE NOT NULL
        )`);

        // Projects Table
        db.run(`CREATE TABLE Projects (
            project_id INTEGER PRIMARY KEY AUTOINCREMENT,
            project_name TEXT NOT NULL,
            description TEXT,
            status TEXT DEFAULT 'Planned',
            guide_id INTEGER,
            FOREIGN KEY (guide_id) REFERENCES ProjectGuides(faculty_id)
        )`);

        // Students Table
        db.run(`CREATE TABLE Students (
            student_id INTEGER PRIMARY KEY AUTOINCREMENT,
            first_name TEXT NOT NULL,
            last_name TEXT NOT NULL,
            email TEXT UNIQUE NOT NULL
        )`);

        // Skills Master List
        db.run(`CREATE TABLE Skills (
            skill_id INTEGER PRIMARY KEY AUTOINCREMENT,
            skill_name TEXT UNIQUE NOT NULL
        )`);

        // Student Skills (Many-to-Many)
        db.run(`CREATE TABLE Student_Skills (
            student_id INTEGER,
            skill_id INTEGER,
            proficiency_level TEXT,
            PRIMARY KEY (student_id, skill_id),
            FOREIGN KEY (student_id) REFERENCES Students(student_id),
            FOREIGN KEY (skill_id) REFERENCES Skills(skill_id)
        )`);

        // Project Allocations
        db.run(`CREATE TABLE ProjectAllocation (
            proj_allocation_id INTEGER PRIMARY KEY AUTOINCREMENT,
            project_id INTEGER,
            student_id INTEGER UNIQUE, -- A student can only have one project
            FOREIGN KEY (project_id) REFERENCES Projects(project_id),
            FOREIGN KEY (student_id) REFERENCES Students(student_id)
        )`);

        console.log("Tables created.");

        // 3. Insert Sample Data (DML)

        // Add Guides
        const stmtGuide = db.prepare("INSERT INTO ProjectGuides (faculty_id, first_name, last_name, email) VALUES (?, ?, ?, ?)");
        stmtGuide.run(501, 'Alice', 'Wong', 'alice@uni.edu');
        stmtGuide.run(502, 'Bob', 'Green', 'bob@uni.edu');
        stmtGuide.finalize();

        // Add Projects
        const stmtProj = db.prepare("INSERT INTO Projects (project_name, description, status, guide_id) VALUES (?, ?, ?, ?)");
        stmtProj.run('ML Data Analysis', 'Analyze COVID-19 trends using Python.', 'In Progress', 501);
        stmtProj.run('Library Management', 'Web app for library book tracking.', 'Planned', 502);
        stmtProj.finalize();

        // Add Students
        const stmtStu = db.prepare("INSERT INTO Students (first_name, last_name, email) VALUES (?, ?, ?)");
        stmtStu.run('John', 'Doe', 'john@uni.edu');
        stmtStu.run('Jane', 'Smith', 'jane@uni.edu');
        stmtStu.run('Mike', 'Ross', 'mike@uni.edu');
        stmtStu.finalize();

        // Add Skills
        const stmtSkill = db.prepare("INSERT INTO Skills (skill_name) VALUES (?)");
        stmtSkill.run('Python');
        stmtSkill.run('SQL');
        stmtSkill.run('React');
        stmtSkill.finalize();

        // Assign Skills to Students
        // John knows Python (Expert), Jane knows SQL (Intermediate)
        db.run("INSERT INTO Student_Skills (student_id, skill_id, proficiency_level) VALUES (1, 1, 'Expert')");
        db.run("INSERT INTO Student_Skills (student_id, skill_id, proficiency_level) VALUES (2, 2, 'Intermediate')");

        // Assign Students to Projects
        // John is on Project 1
        db.run("INSERT INTO ProjectAllocation (project_id, student_id) VALUES (1, 1)");

        console.log("Sample data inserted. Setup complete.");
    });
}